db/flush_job.cc

```diff
@@ -306,6 +317,272 @@ void FlushJob::Cancel() {
+            kTypeRangeDeletion,    // KV type
+            tombstone.start_key_,  // Key is start key.
+            tombstone.end_key_,    // Value is end key.
+            nullptr,               // KV protection info set as nullptr since it
+                                   // should only be useful for the first add to
+                                   // the original memtable.
+            false,                 // : allow concurrent_memtable_writes_
+                                   // Not seen as necessary for now.
+            nullptr,               // get_post_process_info(m) must be nullptr
+                      // when concurrent_memtable_writes is switched off.
+            nullptr);  // hint, only used when concurrent_memtable_writes_
+                       // is switched on.
+
+        if (!s.ok()) {
+          break;
+        }
+
+        // If new_mem is full, add it to purged_mems, and allocate new
+        // memtable.
+        if (new_mem->ShouldScheduleFlush()) {
+          // Rectify the first sequence number, which (unlike the earliest seq
+          // number) needs to be present in the new memtable.
+          new_mem->SetFirstSequenceNumber(new_first_seqno);
+          purged_mems.push_back(new_mem);
+          new_mem = new MemTable(
+              (cfd_->internal_comparator()), *(cfd_->ioptions()),
+              mutable_cf_options_, cfd_->write_buffer_mgr(),
+              mems_[0]->GetEarliestSequenceNumber(), cfd_->GetID());
+          new_first_seqno = kMaxSequenceNumber;
+        }
+      }
+    }
+
+    // If everything happened smoothly and new_mem contains valid data,
+    // decide if it is flushed to storage or kept in the imm()
+    // memtable list (memory).
+    if (s.ok() && (new_first_seqno != kMaxSequenceNumber)) {
+      // Rectify the first sequence number, which (unlike the earliest seq
+      // number) needs to be present in the new memtable.
+      new_mem->SetFirstSequenceNumber(new_first_seqno);
+
+      // The new_mem is added to the list of immutable memtables
+      // only if it filled at less than 10% capacity (arbitrary heuristic).
+      if (new_mem->ApproximateMemoryUsage() <
+          static_cast<size_t>(
+              ceil(0.1 * mutable_cf_options_.write_buffer_size))) {
+        db_mutex_->Lock();
+        cfd_->imm()->Add(new_mem,
+                         &job_context_->memtables_to_free,
+                         false /* trigger_flush. Adding this memtable will not trigger any flush */);
+        new_mem->Ref();
+        db_mutex_->Unlock();
+      } else {
+        purged_mems.push_back(new_mem);
+      }
+    } else {
+      // In this case, the newly allocated new_mem is empty.
+      assert(new_mem != nullptr);
+      delete new_mem;
+    }
+  }
+
+  // Reacquire the mutex for WriteLevel0 function.
+  db_mutex_->Lock();
+
+  // If mempurge successful, don't write input tables to level0,
+  // but write any full output table to level0.
```

Comment:
I guess the track you're taking here is keeping some data in memory only to maximize the chances that it becomes garbage and never has to be flushed. My inclination is (per column family) to flush all or nothing, for the benefit of write amplification in compaction. When we create an L0 file, it should be as big as possible. This should also make it easier to purge obsolete WAL files in due course.

Doesn't necessarily have to be fixed now, but something to be noted for reconsideration.

Reply:
Agreed, provided we're allowed to go beyond the size of a regular SST file (so that we dont accidentally create 2 sst files, one of them being full, the other one filled at xx% capacity). According to the builder.cc:BuildTable code, it looks like we would indeed create a single large SST file, which would be ideal.

Reply:
Before this comment, i actually thought that would lead to the creation of more than 1 sst file because i was convinced the SST file size was enforced. But after code inspection it looks like we would just write a single big SST file.

Reply:
The SST file size limit is for breaking up sorted runs, primarily so that leveled compaction in L1 and above can operate on parts of sorted runs with reasonable granularity. L0 is special because (AFAIK) we assume each SST file is its own sorted run. :)